cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java

```diff
@@ -54,12 +58,28 @@ protected AbstractProgramRuntimeService(ProgramRunnerFactory programRunnerFactor
   public synchronized RuntimeInfo run(Program program, ProgramOptions options) {
     ProgramRunner runner = programRunnerFactory.create(ProgramRunnerFactory.Type.valueOf(program.getType().name()));
     Preconditions.checkNotNull(runner, "Fail to get ProgramRunner for type " + program.getType());
-    final SimpleRuntimeInfo runtimeInfo = new SimpleRuntimeInfo(runner.run(program, options), program);
-    addRemover(runtimeInfo);
+    final RuntimeInfo runtimeInfo = createRuntimeInfo(runner.run(program, createProgramOptionsWithRunId(options)),
+                                                      program);
+    programStarted(runtimeInfo);
     runtimeInfos.put(runtimeInfo.getType(), runtimeInfo.getController().getRunId(), runtimeInfo);
     return runtimeInfo;
   }
 
+  private ProgramOptions createProgramOptionsWithRunId(ProgramOptions options) {
+    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+    builder.putAll(options.getArguments().asMap());
+    // Generate the new RunId and add it to the system arguments
+    RunId runId = RunIds.generate();
```

Comment:
You are overloading this method to do two things (generation of RunId and copying/update the ProgramOptions). It's better to have this method as:

`addRunId(ProgramOptions options, RunId runId);`

and call it like:

`ProgramOptions opt = addRunId(options, RunId.generate());`